README
Compiling in linux:
-------------------
autoreconf -i (optional)
./configure
make

Running tests:
--------------
make check
and/or
xml-patch -v -f test.xml -p patch.xml

Patch operations described in rfc5261:
--------------------------------------
- multi-select extension (not in the specification):
  <add msel="//foo"><bar/></add> ; adds <bar> onto <foo> anywhere
  <replace msel="*/foo"><bar/></replace> ; replace each <foo> with <bar>
					   under document root element
  <remove msel="//@attr"/> ; removes all "attr" attributes from the document

Simple and stupid auto diff tool:
---------------------------------
- look diff directory for a simple diff generator
- compares elements at each tree level with a simple (trivial) logic
- XPath selector values use only positional constraints

Testing:
--------
xml-diff -f test.xml -t patch.xml [-v] [-b] [-o file]

option -b does an additonal patch test
option -o file stores diff result


Install:
--------
- make install, better to use packaging - debian included. Rpm spec included
  which will be generated by running ./configure

License:
--------
GNU LGPL

Legal notice (LEGAL_NOTICE):
----------------------------
clarifies some legal aspects of the software


ChangeLog
Later changes, look debian/changelog


22.10.2007 Jari Urpalainen
	* debian/*:
	- added debian files for building
	* xml_diff.c:
	- if changes in the document root element, return <replace>

31.01.2006 Jari Urpalainen
	* added diff directory with a trivial auto diff algorithm
	* xml_patch.c:
	- CDATA sections within text nodes handled properly
	- added "prepend" option and removed positional selections

07.12.2005 Jari Urpalainen
	* xml_patch.c:
	- some signed/unsigned fixes
	- stupid for loop bug fixed

05.12.2005 Jari Urpalainen
	* xml_patch.c:
	- fixed an improper handling of xmlGetNsList result

23.11.2005 Jari Urpalainen
	* xml_patch_main.c:
	- patched doc saved to file (thanks to Forjan Frederic)
	* xml_patch.c:
	- fixed a change in the spec, when comments and PIs are to be replaced;
	  a bug with escaping '<'
	- added a "msel" selector option to <add>, <replace> & <remove>;
	  so multiple nodes can be located and e.g. several nodes can be deleted
	  with a single request (//@elem). This is NOT specified in the I-D. If nested
	  nodes found in replace & remove, only parents are used

21.10.2005 Jari Urpalainen
	* xml_patch.c:
	- fixed a bug with implicit xml-prefixed namespace
          "http://www.w3.org/XML/1998/namespace"

26.09.2005 Jari Urpalainen
	* xml_patch.c:
	- fixed a bug when replacing a text node with empty content
	- bugfix: recursively replace all ns references

14.09.2005 Jari Urpalainen
	* xml_patch.c:
	- fixed a bug with default ns usage
	- tests for multiple overlapping namespaces
	* xml_patch_main.c:
	- better printing

25.05.2005 Jari Urpalainen
	* Initial build



tests/test_diff.c
/**
 * This is part of an XML patch library.
 *
 * Copyright (C) 2010 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari@urpalainen.fi>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include <stdio.h>
#include <string.h>

#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <errno.h>
#include <assert.h>

#include <glib.h>

#include <libxml/tree.h>

#include "xml_diff.h"
#include "xml_patch.h"

static struct state_s {
    char file_from[PATH_MAX];
    char file_to[PATH_MAX];
    const char *directory;
    xmlDoc *from, *to, *diff;
    int dump, opt;
} state[1];

/** patching test, don't care about request namespaces */
static int patch(xmlDoc *doc, xmlNode *node)
{
    int rc = 0;

    if (!strcmp((char *) node->name, "add"))
	rc = xml_patch_add(doc, node);
    else if (!strcmp((char *) node->name, "replace"))
	rc = xml_patch_replace(doc, node);
    else if (!strcmp((char *) node->name, "remove"))
	rc = xml_patch_remove(doc, node);
    else
	rc = -1;

    return rc;
}

static void open_files(struct state_s *state, const char *sz1, const char *sz2)
{
    snprintf(state->file_from, sizeof(state->file_from), "%s/%s",
				state->directory, sz1);
    snprintf(state->file_to, sizeof(state->file_to), "%s/%s",
				state->directory, sz2);

    state->from = xmlParseFile(state->file_from);
    state->to = xmlParseFile(state->file_to);
}

static void setup(void)
{
    state->diff = xmlNewDoc((const xmlChar *) "1.0");

    state->diff->children = xmlNewDocNode(state->diff, NULL,
						(const xmlChar *) "changes",
						NULL);

    xmlSetNs(state->diff->children,
		xmlNewNs(state->diff->children,
			(const xmlChar *) "urn:xml-changes",
			(const xmlChar *) "x"));
}

static void teardown(void)
{
    xmlFreeDoc(state->from);
    xmlFreeDoc(state->to);
    xmlFreeDoc(state->diff);

    xmlCleanupParser();
}

static int patch_test(struct state_s *state)
{
    xmlNode *node = xmlDocGetRootElement(state->diff);
    int rc;

    for (node = node ? node->children : NULL; node; node = node->next) {
	if (node->type == XML_ELEMENT_NODE) {
	    rc = patch(state->from, node);

	    if (rc) {
		fprintf(stderr, "Patch could not be applied with the "
				"produced diff file !\n");
		return -1;
	    }
	}
    }

    rc = xml_exec_diff(state->from, state->to, state->opt,
			state->diff ? state->diff->children : NULL);

    if (rc)
	fprintf(stderr, "Additional patching test FAILED !\n");
    else if (state->dump)
	fprintf(stdout, "Additional patching test succeeded\n");

    return rc;
}

static void diff_test(struct state_s *state, const char *fn,
                      const char *sz1, const char *sz2)
{
    int rc;

    if (state->dump)
	printf("*** starting %s\n", fn);

    setup();
    open_files(state, sz1, sz2);

    /* generating diff */
    rc = xml_exec_diff(state->from, state->to, state->opt,
			state->diff ? state->diff->children : NULL);
    g_assert(rc);

    /* successful diff means rc > 0, checking that patch to original
     * produces equal document with destination */
    if (rc > 0) {
	if (state->dump) {
	    fprintf(stdout, "OLD doc:\n");
	    xmlDocFormatDump(stdout, state->from, 1);
	    fprintf(stdout, "\nNEW doc:\n");
	    xmlDocFormatDump(stdout, state->to, 1);
	    fprintf(stdout, "\nCHANGES:\n");
	    xmlDocDump(stdout, state->diff);
	}

	g_assert(patch_test(state) == 0);
    }

    if (state->dump)
	printf("*** %s ended\n", fn);

    teardown();
}

static void test_12(void)
{
    diff_test(state, __func__, "base1.xml", "base2.xml");
}

static void test_13(void)
{
    diff_test(state, __func__, "base1.xml", "base3.xml");
}

static void test_14(void)
{
    diff_test(state, __func__, "base1.xml", "base4.xml");
}

static void test_15(void)
{
    diff_test(state, __func__, "base1.xml", "base5.xml");
}

static void test_21(void)
{
    diff_test(state, __func__, "base2.xml", "base1.xml");
}

static void test_31(void)
{
    diff_test(state, __func__, "base3.xml", "base1.xml");
}

static void test_41(void)
{
    diff_test(state, __func__, "base4.xml", "base1.xml");
}

static void test_51(void)
{
    diff_test(state, __func__, "base5.xml", "base1.xml");
}

static void test_from_to(void)
{
    diff_test(state, __func__, "from.xml", "to.xml");
}

static void test_to_from(void)
{
    diff_test(state, __func__, "to.xml", "from.xml");
}

static void test_n01(void)
{
    diff_test(state, __func__, "basen.xml", "basen1.xml");
}

static void test_n10(void)
{
    diff_test(state, __func__, "basen1.xml", "basen.xml");
}

static void test_n02(void)
{
    diff_test(state, __func__, "basen.xml", "basen2.xml");
}

static void test_n20(void)
{
    diff_test(state, __func__, "basen2.xml", "basen.xml");
}

int main (int argc, char *argv[])
{
    int ret;

    static struct option const opt_tbl[] =
    {
	{ "directory",	required_argument, NULL, 'd' },
	{ "no_ws",	no_argument, NULL, 'w' },
	{ "short",	no_argument, NULL, 's' },
	{ "debug",	no_argument, NULL, 'g' },
	{ "help",	no_argument, NULL, 'h' },
	{ NULL, 0, NULL, 0 }
    };

    g_test_init(&argc, &argv, NULL);

    state->directory = "../diff/tests";

    while (ret = getopt_long(argc, argv, "hogwd:", opt_tbl, NULL), ret != -1)
    switch (ret) {
    case 'g':
	state->dump = TRUE;
	break;

    case 'd':
	state->directory = optarg;
	break;

    case 'w':
	state->opt = XML_DIFF_NO_WHITESPACE;
	break;

    case 'o':
	state->opt = XML_DIFF_SIZE_OPTIMIZE;
	break;

    case 'h':
    default:
	printf("test-diff [OPTIONS]\n\n"
		"OPTIONS:\n"
		"  -g dump documents\n"
		"  -l dump available test names, no actual runs\n"
		"  -d (--directory) where all test files exist (%s)\n"
		"  --help lists g_test help\n", state->directory
	       );
	return EXIT_SUCCESS;
    }

    g_test_add_func("/patch/test_1_to_2", test_12);
    g_test_add_func("/patch/test_1_to_3", test_13);
    g_test_add_func("/patch/test_1_to_4", test_14);
    g_test_add_func("/patch/test_1_to_5", test_15);
    g_test_add_func("/patch/test_2_to_1", test_21);
    g_test_add_func("/patch/test_3_to_1", test_31);
    g_test_add_func("/patch/test_4_to_1", test_41);
    g_test_add_func("/patch/test_5_to_1", test_51);
    g_test_add_func("/patch/test_from_to", test_from_to);
    g_test_add_func("/patch/test_to_from", test_to_from);
    g_test_add_func("/patch/test_n_0_to_1", test_n01);
    g_test_add_func("/patch/test_n_0_to_2", test_n02);
    g_test_add_func("/patch/test_n_2_to_0", test_n10);
    g_test_add_func("/patch/test_n_1_to_0", test_n20);

    ret = g_test_run();

    return ret;
}



tests/xml_diff_main.c
/**
 * This is part of an XML patch/diff library.
 *
 * Copyright (C) 2005 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include "config.h"

#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <assert.h>
#include <signal.h>
#include <getopt.h>
#include <ctype.h>
#include <unistd.h>
#include <sysexits.h>

#include <glib.h>

#include <libxml/tree.h>

#include "xml_diff.h"
#include "xml_patch.h"


static void usage()
{
    printf("xml-diff [OPTIONS] -f <from> -t <to>\n\n"
	   "OPTIONS:\n"
	   "  -v == verbose output\n"
	   "  -s == try to shorten requests\n"
	   "  -b == bug testing with patch\n" \
	   "  -o <file> == dumb diff output to file\n\n"
	   "returns 0 if documents don't differ, 1 if documents differ and "
	   "76 if diff generation failed\n");
}

/** patching, don't care about request namespaces */
static int patch(xmlDoc *doc, xmlNode *node)
{
    int rc = 0;

    if (!strcmp((char *) node->name, "add"))
	rc = xml_patch_add(doc, node);
    else if (!strcmp((char *) node->name, "replace"))
	rc = xml_patch_replace(doc, node);
    else if (!strcmp((char *) node->name, "remove"))
	rc = xml_patch_remove(doc, node);
    else
	rc = -1;

    return rc;
}

int main(int argc, char *argv[])
{
    int opt, rc = EXIT_SUCCESS, dump = 0, optimize = 0, bugs = 0;
    const char *doc_from = NULL, *doc_to = NULL, *output = NULL;
    xmlDoc *from, *to, *diff;

    static const struct option opts[] = {
	{ "help", no_argument, NULL, 'h' },
	{ "version", no_argument, NULL, 1 },
	{ NULL, 0, NULL, 0 } };

    while ((opt = getopt_long(argc, argv, "vsbf:t:o:h", opts, NULL)) != -1) {
	switch (opt) {
	case 1:
	    printf("xml-diff version: %s\n", VERSION);
	    return EXIT_SUCCESS;

	case 'h':
	    usage();
	    return EXIT_SUCCESS;

	case '?':
	    usage();
	    return EXIT_FAILURE;

	case 'v':
	    dump = 1;
	    break;

	case 's':
	    optimize = 1;
	    break;

	case 'b':
	    bugs = 1;
	    break;

	case 'f':
	    doc_from = optarg;
	    break;

	case 't':
	    doc_to = optarg;
	    break;

	case 'o':
	    output = optarg;
	    break;
	}
    }
    if (doc_to == NULL || doc_from == NULL) {
	usage();
	return EXIT_FAILURE;
    }
    diff = xmlNewDoc((const xmlChar*) "1.0");

    diff->children = xmlNewDocNode(diff, NULL,
					(const xmlChar*)"changes", NULL);
    xmlSetNs(diff->children,
	     xmlNewNs(diff->children,
			(const xmlChar*) "urn:xml-changes",
			(const xmlChar*) "x"));

    from = xmlParseFile(doc_from);
    to = xmlParseFile(doc_to);

    rc = xml_exec_diff(from, to, optimize, diff->children);

    if (dump && rc >= 0) {
	fprintf(stdout, "OLD doc:\n");
	xmlDocFormatDump(stdout, from, 1);

	if (rc > 0) {
	    fprintf(stdout, "\nNEW doc:\n");
	    xmlDocFormatDump(stdout, to, 1);
	    fprintf(stdout, "\nCHANGES:\n");
	    xmlDocDump(stdout, diff);
	} else if (rc == 0) {
	    fprintf(stdout, "\nEqual documents within xml-diff scope\n");
	}
    }
    if (rc < 0)
	fprintf(stderr, "Diff could not be generated\n");

    if (output && rc > 0)
	xmlSaveFile(output, diff);

    if (rc > 0 && bugs) {
	xmlNode *node = xmlDocGetRootElement(diff);

	for (node = node ? node->children : NULL;
		node; node = node->next) {
	    if (node->type != XML_ELEMENT_NODE)
		continue;

	    rc = patch(from, node);

	    if (rc) {
		fprintf(stderr, "Patch could not be applied with the "
					"produced diff file !\n");
		break;
	    }
	}

	/* check still that patched from == to */
	if (!rc) {
	    rc = xml_exec_diff(from, to, optimize,
				diff ? diff->children : NULL);

	    if (rc)
		fprintf(stderr, "Additional patching test FAILED !\n");
	    else
		fprintf(stdout, "Additional patching test succeeded\n");
	}
    }
    xmlFreeDoc(from);
    xmlFreeDoc(to);
    xmlFreeDoc(diff);

    xmlCleanupParser();

    return rc == 0 ? EXIT_SUCCESS : rc > 0 ? EXIT_FAILURE : EX_PROTOCOL;
}


src/xml_patch.h
/**
 * This is part of an XML patch library.
 *
 * Copyright (C) 2005 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#ifndef XML_PATCH_H
#define XML_PATCH_H

#ifdef __cplusplus
extern "C" {
#endif

int xml_patch_add(xmlDoc *doc, xmlNode *node);
int xml_patch_remove(xmlDoc *doc, xmlNode *node);
int xml_patch_replace(xmlDoc *doc, xmlNode *node);

#ifdef __cplusplus
}
#endif

#endif /* XML_PATCH_H */


diff/xml_diff.h
/**
 * This is part of an XML patch/diff library.
 *
 * Copyright (C) 2005 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#ifndef XML_DIFF_H
#define XML_DIFF_H

#ifdef __cplusplus
extern "C" {
#endif

#define XML_DIFF_WHITESPACE	0
#define XML_DIFF_NO_WHITESPACE	1
#define XML_DIFF_SIZE_OPTIMIZE	2

int xml_exec_diff(const xmlDoc *prev,	/* old doc */
                  const xmlDoc *new,	/* new doc */
                  int fShort,		/* try to shorten payload */
                  xmlNode *node);	/* a node where patches are appended */

#ifdef __cplusplus
}
#endif

#endif /* XML_DIFF_H */
